engine/src/main/java/org/pentaho/di/trans/steps/excelinput/poi/PoiWorkbook.java

```diff
@@ -0,0 +1,165 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.excelinput.poi;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.apache.commons.vfs2.FileObject;
+import org.apache.commons.vfs2.provider.local.LocalFile;
+import org.apache.poi.EncryptedDocumentException;
+import org.apache.poi.openxml4j.opc.OPCPackage;
+import org.apache.poi.poifs.filesystem.POIFSFileSystem;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.logging.KettleLogStore;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.spreadsheet.KSheet;
+import org.pentaho.di.core.spreadsheet.KWorkbook;
+import org.pentaho.di.core.vfs.KettleVFS;
+
+public class PoiWorkbook implements KWorkbook {
+
+  private LogChannelInterface log;
+
+  private Workbook workbook;
+  private String filename;
+  private String encoding;
+  // for PDI-10251 we need direct access to streams
+  private InputStream internalIS;
+  private POIFSFileSystem npoifs;
+  private OPCPackage opcpkg;
+
+  public PoiWorkbook( String filename, String encoding ) throws KettleException {
+    this( filename, encoding, null );
+  }
+
+  public PoiWorkbook( String filename, String encoding, String password ) throws KettleException {
+    this.filename = filename;
+    this.encoding = encoding;
+    this.log = KettleLogStore.getLogChannelInterfaceFactory().create( this );
+    try {
+      FileObject fileObject = KettleVFS.getFileObject( filename );
+      if ( fileObject instanceof LocalFile ) {
+        // This supposedly shaves off a little bit of memory usage by allowing POI to randomly access data in the file
+        //
+        String localFilename = KettleVFS.getFilename( fileObject );
+        File excelFile = new File( localFilename );
+        try {
+          npoifs = new POIFSFileSystem( excelFile );
+          workbook = org.apache.poi.ss.usermodel.WorkbookFactory.create( npoifs );
+        } catch ( Exception ofe ) {
+          try {
+            opcpkg = OPCPackage.open( excelFile );
+            workbook = org.apache.poi.ss.usermodel.WorkbookFactory.create( opcpkg );
+          } catch ( Exception ex ) {
+            workbook = org.apache.poi.ss.usermodel.WorkbookFactory.create( excelFile, password );
+          }
+        }
+      } else {
+        internalIS = KettleVFS.getInputStream( filename );
+        workbook = org.apache.poi.ss.usermodel.WorkbookFactory.create( internalIS, password );
+      }
+    } catch ( EncryptedDocumentException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)